### DISCUSSION TITLE
Error: ENOENT: no such file or directory, open ../snapshot-db_name.json

### DISCUSSION THREAD

**Discussion by Juknum** [mikro-orm/mikro-orm#4307] [Q&A] [answered]
**Describe the bug**
Hi, I would like to run the `npx mikro-orm migration:create --initial` cli command but I'm stuck on a ENOENT error;
> By the way I would like to set up a "base" for my database when I first init my project, is it the right way to do it ? 

**Stack trace**
```bash
mikro-orm migration:create

Create new migration with current schema diff

Options:
  -b, --blank    Create blank migration                                [boolean]
  -i, --initial  Create initial migration                              [boolean]
  -d, --dump     Dumps all queries to console                          [boolean]
  -p, --path     Sets path to directory where to save entities          [string]
  -v, --version  Show version number                                   [boolean]
  -h, --help     Show help                                             [boolean]

[Error: ENOENT: no such file or directory, open '/Users/juknum/Documents/Code/AE/api/src/migrations/.snapshot-ae_test.json'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/juknum/Documents/Code/AE/api/src/migrations/.snapshot-ae_test.json'
}
```

**To Reproduce**
1. Run `npx mikro-orm migration:create --initial`

**Expected behavior**
The migration class file should be created ?

**Additional context**

I'm using Mikro-ORM with NestJS & GraphQL ([GitHub repo](https://github.com/ae-utbm/api/tree/setup-mikrorm)) and I use the following config file :

```ts
// mikro-orm.config.ts
import { MikroORMOptions, IDatabaseDriver, Connection } from '@mikro-orm/core';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
import { Logger } from '@nestjs/common';
import { SqlHighlighter } from '@mikro-orm/sql-highlighter';

const logger = new Logger('MikroORM');

export default (): Partial<MikroORMOptions<IDatabaseDriver<Connection>>> => ({
	type: 'postgresql',
	dbName: process.env.DB_NAME, // 'postgres'
	port: parseInt(process.env.DB_PORT, 10), // 5432
	host: process.env.DB_HOST, // '127.0.0.1'
	user: process.env.DB_USER, // 'root'
	debug: process.env.DEBUG === 'true',
	entities: ['./dist/modules/**/entities/*.entity.js'],
	entitiesTs: ['./src/modules/**/entities/*.entity.ts'],
	highlighter: new SqlHighlighter(),
	migrations: {
		snapshot: false,
		transactional: true,
		path: './dist/database/migrations',
		pathTs: './src/database/migrations',
	},
	logger: logger.log.bind(logger),
	metadataProvider: TsMorphMetadataProvider,
});
```

**Versions**

| Dependency | Version |
| - | - |
| node | `20.0.0` |
| typescript | `^4.9.5` |
| mikro-orm | `^5.7.4` |
| your-driver | ? |

**Comment by B4nan** [comment]
Does it work if you drop the `--initial`? You almost never need this, only in case you already have an existing schema.

> snapshot: false

Just curious, did you add this when trying to get around the error? You should generally keep snapshots enabled. It's indeed weird that it tries to load the snapshot even when it's disabled.

**Comment by Juknum** [comment]
Yup I have the `No changes required, schema is up-to-date` message but I don't have any migration file created.

And yeah I have added the snapshot settings as a workaround but nothing changed :/

**Comment by B4nan** [comment]
Can you try to add this line [here](https://github.com/mikro-orm/mikro-orm/blob/master/packages/migrations/src/Migrator.ts#L376)?

```diff
private async ensureMigrationsDirExists() {
  if (!this.options.migrationsList) {
    await ensureDir(this.absolutePath);
+   await ensureDir(this.snapshotPath);
  }
}
```

**Comment by Juknum** [comment]
I only have the compiled JavaScript inside the `node_modules` directory, so I have written this : 
```diff
// node_modules/@mikro-orm/migrations/Migrator.js

async ensureMigrationsDirExists() {
  if (!this.options.migrationsList) {
    await (0, fs_extra_1.ensureDir)(this.absolutePath);
+   await (0, fs_extra_1.ensureDir)(this.snapshotPath);
  }
}
```

But the error is still there

**Comment by B4nan** [comment]
That's really weird, try to find the place that is causing the error. I checked the code, and all places I see that are trying to read or write the snapshot are both checking the option you have disabled, as well as the file existence itself. Only the `Migrator` class should be working with the snapshot.

**Comment by B4nan** [comment]
Also the fact that it tries to open file `'/Users/juknum/Documents/Code/AE/api/src/migrations/.snapshot-ae_test.json'` is weird - it should respect the migration paths you set to `'./src/database/migrations'`:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/migrations/src/Migrator.ts#L35-L39

**Comment by Juknum** [comment]
Yeah that's weird, I don't think my config does anything, when I add a `console.log(this.options)` I get the following : 
```js
{
  tableName: 'mikro_orm_migrations',
  path: './dist/migrations',
  glob: '!(*.d).{js,ts,cjs}',
  silent: false,
  transactional: true,
  disableForeignKeys: true,
  allOrNothing: true,
  dropTables: true,
  safe: false,
  snapshot: true,
  emit: 'ts',
  fileName: [Function: fileName],
  pathTs: './src/migrations'
}
``` 

Which looks like a default config file, is there anything I am missing ?

**Comment by B4nan** [comment]
Maybe its caused by you defining the config as a callback? It should work, but maybe it breaks the config merging with the defaults?

**Comment by Juknum** [comment]
Nope I've tried to switch from the callback to a plain object and it still logging the "default" config, do I need to specify anything in the `package.json` file about `mikro-orm` for the cli ? 🤔

**Comment by B4nan** [comment]
Yes, I see you skipped that. I think the `useTsNode` option could be enough, that will enable look up of the TS config. The `src/dist` folders are nowadays checked automatically, but only if you are inside ts-node context. The CLI won't register ts-node without you enabling the `useTsNode` flag (and installing ts-node, which I see you already have).

```json
  "name": "our-app",
  "dependencies": { ... },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
}
```

https://mikro-orm.io/docs/installation#setting-up-the-commandline-tool

**Comment by B4nan** [comment]
Yeah, this is the PR that handles the smarter dynamic defaults #3497, and it explicitly says the `useTsNode` is still required.

**Comment by Juknum** [comment]
Okay I know have a more specific error : 
```bash
Error: Cannot find module 'src/database/entities/base.entity'
Require stack:
- /Users/juknum/Documents/Code/AE/api/src/modules/auth/entities/refresh-token.entity.ts
- /Users/juknum/Documents/Code/AE/api/__fake.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1082:15)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/juknum/Documents/Code/AE/api/node_modules/.pnpm/@cspotcode+source-map-support@0.8.1/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:928:27)
    at Module.require (node:internal/modules/cjs/loader:1149:19)
    at require (node:internal/modules/helpers:121:18)
    at Object.<anonymous> (/Users/juknum/Documents/Code/AE/api/src/modules/auth/entities/refresh-token.entity.ts:2:1)
    at Module._compile (node:internal/modules/cjs/loader:1267:14)
    at Module.m._compile (/Users/juknum/Documents/Code/AE/api/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.11.18_typescript@4.9.5/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1321:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/juknum/Documents/Code/AE/api/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.11.18_typescript@4.9.5/node_modules/ts-node/src/index.ts:1621:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/juknum/Documents/Code/AE/api/src/modules/auth/entities/refresh-token.entity.ts',
    '/Users/juknum/Documents/Code/AE/api/__fake.js'
  ]
}
```
But I did found why it was happening aha, it seems that the cli doesn't handle paths from the base dir (relatives one works)

```diff
- import { BaseEntity } from 'src/database/entities/base.entity';
- import { User } from 'src/modules/users/entities/user.entity';
+ import { BaseEntity } from '../../../database/entities/base.entity';
+ import { User } from '../../users/entities/user.entity';
```

**Comment by B4nan** [comment]
> import { BaseEntity } from 'src/database/entities/base.entity';

This is not an absolute path, this is paths mapping - you are apparently using tsconfig-paths, and you need to configure them. The CLI does support that too, as long as you defined them in the root tsconfig.json (or you set the path to the correct tsconfig in your package.json).

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/ConfigurationLoader.ts#L143-L148

**Comment by B4nan** [comment]
That project setup is quite weird, not sure if nest is handling the paths mapping somehow automatically, but I don't see any place where you would configure it. 

I'd rather suggest using relative paths, they work universally, don't depend on a 3rd party package, and will generally save you a lot of troubles.

I will move this into discussions, as it is clearly a configuration issue on your side. Will try to think about how to improve this in v6.

**Comment by Juknum** [comment, marked as answer]
Ok thanks for everything ! I was able to run the command without any issue :)

I had to add the following to my `tsconfig.json` : 
```diff
{
	"compilerOptions": {
		...
+		"paths": {
+			"@/*": [
+				"src/*"
+			]
+		}
	}
}
```

This to the `package.json` file
```diff
{
	"name": "api",
	"version": "1.0.0",
	...
+	"mikro-orm": {
+		"useTsNode": true
+	}

```

**Reply by B4nan** [reply]
I think you can leave the snapshots enabled, or does it fail somehow with the fixed configs? Also the `configPaths` shouldn't be necessary (with those specific values), would be great if you can try to remove those two bits and confirm it still works, as it should.

**Reply by Juknum** [reply]
Uh yeah I can try but how do I remove the initial migration ? I have the error : 
`Error: Initial migration cannot be created, as some migrations already exist` (using the `--initial` flag)

When I remove the flag I get the following error : 
`Cannot find module '/Users/juknum/Documents/Code/AE/api/src/database/migrations/.snapshot-ae_test.json'`
Which is present inside the migrations directory but as a folder and not a json file ://

<p align=center><img width="223" alt="Capture d’écran 2023-05-02 à 17 14 34" src="https://user-images.githubusercontent.com/49886317/235709193-f7a0ee54-379e-449d-81e7-fb7769bcf349.png"></p>

**Reply by B4nan** [reply]
> Uh yeah I can try but how do I remove the initial migration ? I have the error :

Just remove the generated migration file (together with the wrongly created snapshot folder). And you might have to wipe the migrations table from your schema too.

> Which is present inside the migrations directory but as a folder and not a json file ://

I think this comes from the added line with `await ensureDir(this.snapshotPath)`, it was a wrong suggestion as the path points to the file, not to the folder where it sits.

**Reply by Juknum** [reply]
>Just remove the generated migration file (together with the wrongly created snapshot folder). And you might have to wipe the migrations table from your schema too.

Yeah it worked, I had to drop the schema using `npx mikro-orm schema:drop --dump ` though

>  it was a wrong suggestion as the path points to the file, not to the folder where it sits.   

Yup that was that

I'll edit my message so the selected answer is more correct

**Reply by B4nan** [reply]
Ok great, thanks for the confirmation!

I will think about the `useTsNode` stuff in v6, it would be great to remove the need for that too, I know this can be quite confusing, and while this is documented on the very first page in the docs, it can be easy to miss.
